File: src/backend/models/places.rs
use serde::{Deserialize, Serialize};

use super::MediaResponse;

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct PlacesListRequest {
    pub cursor: Option<String>,
    pub limit: i64,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct PlaceGetRequest {
    pub place_id: String,
    pub cursor: Option<String>,
    pub limit: i64,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct PlaceSummary {
    pub place_id: String,
    pub city: String,
    pub state: Option<String>,
    pub country: String,
    pub media_count: i64,
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct PlacesListResponse {
    pub places: Vec<PlaceSummary>,
    pub next_cursor: Option<String>,
    pub has_more: bool,
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct PlaceGetResponse {
    pub place: PlaceSummary,
    pub media: Vec<MediaResponse>,
    pub next_cursor: Option<String>,
    pub has_more: bool,
}

File: src/frontend/api/places.ts
import { apiClient } from './client'
import type { Media } from './types'

export interface PlaceSummary {
  placeId: string
  city: string
  state: string | null
  country: string
  mediaCount: number
}

export interface PlacesListRequest {
  cursor: string | null
  limit: number
}

export interface PlacesListResponse {
  places: PlaceSummary[]
  nextCursor: string | null
  hasMore: boolean
}

export interface PlaceGetRequest {
  placeId: string
  cursor: string | null
  limit: number
}

export interface PlaceGetResponse {
  place: PlaceSummary
  media: Media[]
  nextCursor: string | null
  hasMore: boolean
}

export const placesApi = {
  list: async (request: PlacesListRequest): Promise<PlacesListResponse> => {
    const response = await apiClient.post<PlacesListResponse>('/places/list', request)
    return response.data
  },

  get: async (request: PlaceGetRequest): Promise<PlaceGetResponse> => {
    const response = await apiClient.post<PlaceGetResponse>('/places/get', request)
    return response.data
  },

  getThumbnail: async (placeId: string): Promise<string> =>
    `/api/v1/places/${encodeURIComponent(placeId)}/thumbnail`,
}
